File: programacao-dinamica/python/min_cost_climbing_stairs.py
def min_cost_climbing_stairs(cost):
    memo = [-1] * (len(cost) + 1)  # Inicializa o array de memoização

    return min_cost_climbing_stairs_util(len(cost), cost, memo)

def min_cost_climbing_stairs_util(n, cost, memo):
    if n == 0 or n == 1:
        return 0

    if memo[n] != -1:
        return memo[n]

    memo[n] = min(
        min_cost_climbing_stairs_util(n - 1, cost, memo) + cost[n - 1],
        min_cost_climbing_stairs_util(n - 2, cost, memo) + cost[n - 2]
    )

    return memo[n]

# Casos de teste
cost1 = [10, 15, 20]
print(min_cost_climbing_stairs(cost1))  # Saída esperada: 15

cost2 = [1, 100, 1, 1, 1, 100, 1, 1, 100, 1]
print(min_cost_climbing_stairs(cost2))  # Saída esperada: 6

File: programacao-dinamica/javascript/mincost_climbing_stairs.js
/**
 * @param {number[]} cost - Um array contendo o custo de cada degrau para subir
 * @return {number} - O custo mínimo para subir os degraus
 */
var minCostClimbingStairs = function(cost) {
    const memo = Array(cost.length + 1).fill(-1); // Inicializa o array de memoização

    return minCostClimbingStairsUtil(cost.length, cost, memo);
};

/**
 * Função auxiliar para calcular o custo mínimo de subir os degraus utilizando memoização.
 * @param {number} n - O número de degraus a subir
 * @param {number[]} cost - Um array contendo o custo de cada degrau para subir
 * @param {number[]} memo - Array de memoização para armazenar resultados intermediários
 * @return {number} - O custo mínimo para subir os degraus até o degrau n
 */
var minCostClimbingStairsUtil = function(n, cost, memo) {
    if (n === 0 || n === 1) {
        return 0;
    }

    if (memo[n] !== -1) {
        return memo[n];
    }

    memo[n] = Math.min(
        minCostClimbingStairsUtil(n - 1, cost, memo) + cost[n - 1],
        minCostClimbingStairsUtil(n - 2, cost, memo) + cost[n - 2]
    );

    return memo[n];
};

// Caso de teste 1
let cost1 = [10, 15, 20];
console.log(minCostClimbingStairs(cost1));

// Caso de teste 2
let cost2 = [1, 100, 1, 1, 1, 100, 1, 1, 100, 1];
console.log(minCostClimbingStairs(cost2));
